File: Managed/UnrealBowlingUSharp.Managed/BowlingGameComponents/BowlingBallComponent.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using UnrealEngine.Runtime;
using UnrealEngine.Engine;
using UnrealEngine.Slate;
using UnrealEngine;
using UnrealEngine.GameplayTasks;
using UnrealEngine.SlateCore;
using UnrealEngine.NavigationSystem;

namespace HelloUSharp
{
    [UClass, Blueprintable, BlueprintType]
    public class BowlingBallComponent : UActorComponent
    {
        #region IgnoreProperties
        [UPropertyIngore]
        public AActor MyOwner
        {
            get
            {
                if (_owner == null)
                    _owner = GetOwner();

                return _owner;
            }
        }
        private AActor _owner = null;

        [UPropertyIngore]
        private FVector myPos
        {
            get { return MyOwner.GetActorLocation(); }
        }

        [UPropertyIngore]
        public UStaticMeshComponent MyMeshComponent { get; set; }

        [UPropertyIngore]
        public UAudioComponent MyAudioSourceComponent { get; set; }

        [UPropertyIngore]
        protected BowlGameMasterComponent gamemaster => BowlGameMasterComponent.GetInstance(MyOwner);
        [UPropertyIngore]
        protected BowlGameModeComponent gamemode => BowlGameModeComponent.GetInstance(MyOwner);
        #endregion

        #region Fields
        private FHitResult myHit;
        private FVector MyStartLocation;
        private FRotator MyStartRotation;
        #endregion

        #region MyUProperties
        [UProperty, EditAnywhere, BlueprintReadWrite, Category("Bowling")]
        public FVector LaunchVelocity { get; set; }

        [UProperty, EditDefaultsOnly, BlueprintReadWrite, Category("Initialization")]
        public USoundBase BallRollingSound { get; set; }

        [UProperty, EditDefaultsOnly, BlueprintReadWrite, Category("Initialization")]
        public USoundBase BallNudgeSound { get; set; }
        #endregion

        #region Overrides
        public override void Initialize(FObjectInitializer initializer)
        {
            //base.Initialize();
        }

        public override void BeginPlay()
        {
            //base.ReceiveBeginPlay_Implementation();
            //LaunchBall();
            gamemaster.BowlTurnIsFinished += BowlTurnIsFinished;
            gamemaster.BowlNewTurnIsReady += NewTurnIsReady;
            //gamemaster.BowlTurnIsFinishedTest.Bind(BowlTurnIsFinished);
            gamemaster.OnBallLaunch += LaunchBall;
            gamemaster.OnNudgeBallLeft += NudgeBallLeft;
            gamemaster.OnNudgeBallRight += NudgeBallRight;

            MyStartLocation = MyOwner.GetActorLocation();
            MyStartRotation = MyOwner.GetActorRotation();
        }

        protected override void ReceiveTick_Implementation(float DeltaSeconds)
        {
            //base.ReceiveTick_Implementation(DeltaSeconds);
            //SetActorLocation(myPos + new FVector(DefaultMoveSpeed, 0, 0), false, out myHit, false);
        }

        public override void EndPlay(EEndPlayReason endPlayReason)
        {
            if (gamemaster != null)
            {
                gamemaster.BowlTurnIsFinished -= BowlTurnIsFinished;
                gamemaster.BowlNewTurnIsReady -= NewTurnIsReady;
                //gamemaster.BowlTurnIsFinishedTest.Unbind(BowlTurnIsFinished);
                gamemaster.OnBallLaunch -= LaunchBall;
                gamemaster.OnNudgeBallLeft -= NudgeBallLeft;
                gamemaster.OnNudgeBallRight -= NudgeBallRight;
            }
        }
        #endregion

        #region Handlers
        void NewTurnIsReady(EBowlAction _action)
        {
            if (MyMeshComponent == null) return;

            MyOwner.SetActorLocation(
                MyStartLocation, false, out myHit, false
                );
            MyOwner.SetActorRotation(
                MyStartRotation, false
                );

            MyMeshComponent.SetSimulatePhysics(false);
            MyMeshComponent.SetSimulatePhysics(true);
        }

        //Only Public Because I was testing Delegate Binding
        [UFunction, BlueprintCallable]
        public void BowlTurnIsFinished()
        {
            //MyOwner.PrintString("BowlTurnIsFinishedd", FLinearColor.Green);
        }

        void LaunchBall(FVector launchVelocity, BowlingBallComponent bowlingBall)
        {
            if (MyMeshComponent == null)
            {
                MyOwner.PrintString("Please Assign A mesh component to the uproperty", FLinearColor.OrangeRed);
            }
            else if (MyAudioSourceComponent == null)
            {
                MyOwner.PrintString("Please Assign an audio component to the uproperty", FLinearColor.OrangeRed);
            }
            else if (BallRollingSound == null)
            {
                MyOwner.PrintString("Please Assign a sound clip to the ball rolling sound uproperty", FLinearColor.OrangeRed);
            }
            else
            {
                MyMeshComponent.AddImpulse(launchVelocity, MyMeshComponent.GetAttachSocketName(), true);
                MyAudioSourceComponent.Sound = BallRollingSound;
                MyAudioSourceComponent.Play();
            }
        }

        void NudgeBallLeft(float famount)
        {
            FHitResult _hit;
            MyOwner.SetActorLocation(
                MyOwner.GetActorLocation() +
                new FVector(0, famount, 0), false, out _hit, false);

            if (BallNudgeSound != null)
            {
                MyAudioSourceComponent.Sound = BallNudgeSound;
                MyAudioSourceComponent.Play();
            }
        }

        void NudgeBallRight(float famount)
        {
            FHitResult _hit;
            MyOwner.SetActorLocation(
                MyOwner.GetActorLocation() +
                new FVector(0, famount, 0), false, out _hit, false);

            if (BallNudgeSound != null)
            {
                MyAudioSourceComponent.Sound = BallNudgeSound;
                MyAudioSourceComponent.Play();
            }
        }
        #endregion

        #region Initialization
        [UFunction, BlueprintCallable]
        public void MyBeginPlayInitializer(UStaticMeshComponent _mymeshcomponent, UAudioComponent _myaudiosourcecomponent)
        {
            MyMeshComponent = _mymeshcomponent;
            MyAudioSourceComponent = _myaudiosourcecomponent;
        }
        #endregion
    }
}

File: Source/UnrealBowlingUSharp/Private/BowlingBallComponent.cpp
// Fill out your copyright notice in the Description page of Project Settings.


#include "Public/BowlingBallComponent.h"
#include "Kismet/GameplayStatics.h"
#include "GameFramework/GameModeBase.h"
#include "BowlGameMasterComponent.h"
#include "BowlGameModeComponent.h"
#include "Sound/SoundBase.h"
#include "Components/StaticMeshComponent.h"
#include "Components/AudioComponent.h"

#pragma region Component Getters
UBowlGameMasterComponent* UBowlingBallComponent::GetGameMaster()
{
	if (bowlGameMaster == nullptr)
	{
		auto _gamemode = UGameplayStatics::GetGameMode(this);
		if (_gamemode != nullptr)
		{
			bowlGameMaster = Cast<UBowlGameMasterComponent>(_gamemode->GetComponentByClass(UBowlGameMasterComponent::StaticClass()));
		}
	}
	return bowlGameMaster;
}

UBowlGameModeComponent* UBowlingBallComponent::GetBowlGameMode()
{
	if (bowlGameMode == nullptr)
	{
		auto _gamemode = UGameplayStatics::GetGameMode(this);
		if (_gamemode != nullptr)
		{
			bowlGameMode = Cast<UBowlGameModeComponent>(_gamemode->GetComponentByClass(UBowlGameModeComponent::StaticClass()));
		}
	}
	return bowlGameMode;
}
#pragma endregion

#pragma region InitAndOverrides
void UBowlingBallComponent::MyBeginPlayInitializer(UStaticMeshComponent* _mymeshcomponent,
	UAudioComponent* _myaudiosourcecomponent)
{
	this->MyMeshComponent = _mymeshcomponent;
	this->MyAudioSourceComponent = _myaudiosourcecomponent;
}

// Sets default values for this component's properties
UBowlingBallComponent::UBowlingBallComponent()
{
	// Set this component to be initialized when the game starts, and to be ticked every frame.  You can turn these features
	// off to improve performance if you don't need them.
	PrimaryComponentTick.bCanEverTick = true;

	// ...
}


// Called when the game starts
void UBowlingBallComponent::BeginPlay()
{
	Super::BeginPlay();

	auto _gamemaster = GetGameMaster();
	if(ensure(_gamemaster != nullptr) == false)
	{
		return;
	}

	if (ensure(_gamemaster->BowlTurnIsFinished.IsAlreadyBound(this, &UBowlingBallComponent::BowlTurnIsFinished) == false))
	{
		_gamemaster->BowlTurnIsFinished.AddDynamic(this, &UBowlingBallComponent::BowlTurnIsFinished);
	}
	if (ensure(_gamemaster->BowlNewTurnIsReady.IsAlreadyBound(this, &UBowlingBallComponent::NewTurnIsReady) == false))
	{
		_gamemaster->BowlNewTurnIsReady.AddDynamic(this, &UBowlingBallComponent::NewTurnIsReady);
	}
	if (ensure(_gamemaster->OnBallLaunch.IsAlreadyBound(this, &UBowlingBallComponent::LaunchBall) == false))
	{
		if (bLaunchBallThroughBlueprints == false) {
			_gamemaster->OnBallLaunch.AddDynamic(this, &UBowlingBallComponent::LaunchBall);
		}
	}
	if (ensure(_gamemaster->OnNudgeBallLeft.IsAlreadyBound(this, &UBowlingBallComponent::NudgeBallLeft) == false))
	{
		_gamemaster->OnNudgeBallLeft.AddDynamic(this, &UBowlingBallComponent::NudgeBallLeft);
	}
	if (ensure(_gamemaster->OnNudgeBallRight.IsAlreadyBound(this, &UBowlingBallComponent::NudgeBallRight) == false))
	{
		_gamemaster->OnNudgeBallRight.AddDynamic(this, &UBowlingBallComponent::NudgeBallRight);
	}

	MyStartLocation = GetOwner()->GetActorLocation();
	MyStartRotation = GetOwner()->GetActorRotation();
}

// Called every frame
void UBowlingBallComponent::TickComponent(float DeltaTime, ELevelTick TickType, FActorComponentTickFunction* ThisTickFunction)
{
	Super::TickComponent(DeltaTime, TickType, ThisTickFunction);

	// ...
}
#pragma endregion

#pragma region Handlers
void UBowlingBallComponent::NewTurnIsReady(EBowlAction _action)
{
	if (MyMeshComponent == nullptr) return;

	GetOwner()->SetActorLocation(MyStartLocation, false);
	GetOwner()->SetActorRotation(MyStartRotation);

	MyMeshComponent->SetSimulatePhysics(false);
	MyMeshComponent->SetSimulatePhysics(true);
}

void UBowlingBallComponent::BowlTurnIsFinished()
{

}

void UBowlingBallComponent::LaunchBall(FVector _launchVelocity, UBowlingBallComponent* bowlingBall)
{
	if (bLaunchBallThroughBlueprints) return;

	if (MyMeshComponent == nullptr)
	{
		UE_LOG(LogTemp, Warning, TEXT("Please Assign A mesh component to the uproperty"));
		return;
	}
	else if (MyAudioSourceComponent == nullptr)
	{
		UE_LOG(LogTemp, Warning, TEXT("Please Assign an audio component to the uproperty"));
		return;
	}
	else if (BallRollingSound == nullptr)
	{
		UE_LOG(LogTemp, Warning, TEXT("Please Assign a sound clip to the ball rolling sound uproperty"));
		return;
	}
	else
	{
		//UE_LOG(LogTemp, Warning, TEXT("LaunchBall: Launch Velocity X: %f. Launch Velocity Y: %f"), _launchVelocity.X, _launchVelocity.Y);
		MyMeshComponent->AddImpulse(LaunchVelocity, MyMeshComponent->GetAttachSocketName(), true);
		MyAudioSourceComponent->Sound = BallRollingSound;
		MyAudioSourceComponent->Play();
	}
}

void UBowlingBallComponent::NudgeBallLeft(float famount)
{
	GetOwner()->SetActorLocation(
		GetOwner()->GetActorLocation() + FVector(0, famount, 0),
		false
	);
}

void UBowlingBallComponent::NudgeBallRight(float famount)
{
	GetOwner()->SetActorLocation(
		GetOwner()->GetActorLocation() + FVector(0, famount, 0),
		false
	);
}
#pragma endregion